provider/qsd.py

```diff
@@ -0,0 +1,473 @@
+        img["format"].update(json.loads(params.get("qsd_image_format", "{}")))
+        img["filter"].update(json.loads(params.get("qsd_image_filter", "{}")))
+
+        img["export"]["id"] = "id_" + name
+        img["export"]["node-name"] = img["filter"]["node-name"] if \
+            img["filter"]["driver"] else img["format"]["node-name"]
+        img["export"].update(json.loads(params.get("qsd_image_export", "{}")))
+
+        if img["export"]["type"] == "nbd":
+            # The name is necessary. empty value to simply setting in nbd client
+            img["export"]["name"] = ""
+            addr = json.loads(
+                params.get("qsd_image_export_nbd", '{"type":"unix"'))
+
+            img[addr["type"]].update(addr)
+            if addr["type"] == "unix":
+                if not img[addr["type"]]["path"]:
+                    img[addr["type"]]["path"] = "%s/%s_nbd_%s.sock" % (
+                        self.basedir, self.name, name)
+            img["nbd-server"]["addr"].update(img[addr["type"]])
+
+        elif img["export"]["type"] == "vhost-user-blk":
+            img["unix"]["path"] = "%s/%s_vhost_user_%s.sock" % (
+                self.basedir, self.name, name)
+            img["export"]["addr"] = img["unix"]
+        else:
+            raise QsdError("Unknown export type %s " % img["export"]["type"])
+
+        # Prepare image
+        image_created = False
+
+        if name in params.get("images").split():
+            if params.get("force_create_image") == "yes":
+                image_created = True
+            elif params.get(
+                    "create_image") == "yes" and not storage.file_exists(
+                    img["protocol"]["filename"]):
+                image_created = True
+
+        if image_created:
+            LOG_JOB.info("QSD skip to create image %s ", name)
+        else:
+            LOG_JOB.info("QSD ready to create image %s", name)
+            obj = qemu_storage.QemuImg(params, data_dir.get_data_dir(), name)
+            _, result = obj.create(params)
+            if result.exit_status != 0:
+                raise QsdError("Failed create image %s " % name)
+            img["image_object"] = obj
+
+        self.images.update({name: img})
+        return img
+
+    def has_option(self, option):
+        """
+        :param option: Desired option
+        :return: Is the desired option supported by current qemu?
+        """
+        return bool(re.search(r"%s" % option, self.__qsd_help,
+                              re.MULTILINE))
+
+    def _probe_capabilities(self):
+        """ Probe capabilities. """
+
+        if self.has_option('--daemonize'):
+            LOG_JOB.info("--daemonize")
+            self.caps.set_flag(Flags.DAEMONIZE)
+        if self.has_option('--pidfile'):
+            LOG_JOB.info("--pidfile")
+            self.caps.set_flag(Flags.PIDFILE)
+
+    def get_pid(self):
+        """ Get QSD pid"""
+        if self.daemon_process:
```

Comment:
refer to line370 line 372, if self.daemonize, self.daemon_process should be None? 
ie. it's possible to set None to self._daemon_process, at that time, use popen to create a daemonized process(line370)

Reply:
updated